### DISCUSSION TITLE
Fastlane issues

### DISCUSSION THREAD

**Discussion by IzzySoft** [FossifyOrg/General-Discussion#536] [General]
Monthly check in my repo yelled that

```
! metadata/org.fossify.phone/pt/short_description.txt has 89 chars
! metadata/org.fossify.phone/en-US/changelogs/4.txt has 915 chars
```

Shortdesc is limited to max 80 chars, per-release changelogs to max 500 chars. Both will be cut at exactly that position, even if it's mid-word. The latter certainly did not come in via Weblate I guess :wink:

**Comment by naveensingh** [comment]
Thanks! I fixed the short description on Weblate. Max length checks are already enforced.

> The latter certainly did not come in via Weblate I guess 😉

Updated! I didn't know about the 500 character limit, will keep it short next time :)

**Comment by IzzySoft** [comment]
Thanks a lot! Just one question as we're talking about Fastlane: currently I take your fulldesc the way F-Droid does, replacing all newlines with `<br>` – with the exception of Gallery, which I manually edited to be "proper HTML". We spoke about this already in the past but postponed it then. Do you think you want to go with the nicer looking HTML formatting, or stick with the current formatting?

**Comment by naveensingh** [comment]
Actually it's still on my TODO list, I just never got around to it. 

You say Google Play will strip down the html tags? I would like to use a format that works everywhere out of the box so how about something like this:

<details>

<summary>Click to expand:</summary>


```
Empower your calls, safeguard your data. Fossify Phone redefines the mobile app experience with unmatched privacy and efficiency. Free from ads and intrusive permissions, it's designed for seamless and secure everyday communication.
<br>
<br>
<b>📱 YOUR PRIVACY, OUR PRIORITY:</b><br>
Welcome to the Fossify Phone App, where your digital privacy is paramount. Switch to a mobile experience that respects your data, ensuring your personal information remains secure and private.
<br>
<br>
<b>🚀 SEAMLESS PERFORMANCE:</b><br>
The Fossify Phone App offers a fluid and responsive mobile interface, enhancing your phone's performance while safeguarding your privacy. Experience a lag-free, smooth user experience, optimized for efficiency and speed.
<br>
<br>
<b>🌐 OPEN-SOURCE ASSURANCE:</b><br>
With the Fossify Phone App, transparency is at your fingertips. Built on an open-source foundation, our app allows you to review our code on GitHub, fostering trust and a community committed to privacy.
<br>
<br>
<b>🖼️ TAILOR-MADE CUSTOMIZATION:</b><br>
Customize your mobile experience with the Fossify Phone App. Adjust your app settings for a personalized interface, from thematic designs to functional preferences. Enjoy a user interface that's intuitive and uniquely yours.
<br>
<br>
<b>🔋 EFFICIENT RESOURCE MANAGEMENT:</b><br>
The Fossify Phone App is designed for optimal resource usage, contributing to extended battery life. It's light on your phone's resources, ensuring your device runs efficiently with minimized battery drain.
<br>
<br>
Download the Fossify Phone App now and step into a mobile world where privacy seamlessly blends with functionality. Your journey towards a safer, personalized mobile experience starts here.
<br>
<br>
Explore more Fossify apps: https://www.fossify.org<br>
Open-Source Code: https://www.github.com/FossifyOrg<br>
Join the community on Reddit: https://www.reddit.com/r/Fossify<br>
Connect on Telegram: https://t.me/Fossify<br>
```

</details>


I edited the html on the webpage to draw a comparison between `<dl>`, `<dt>` formatting used by the Gallery on your repo and `<br>`, `<b>` formatting I suggested above:  


<img src="https://github.com/FossifyOrg/Phone/assets/36371707/bfd73353-9894-4f03-8a42-f4e2824fa86e" width=400/><img src="https://github.com/FossifyOrg/Phone/assets/36371707/719a35e7-9914-40cc-8ee1-002b028d7a7b" width=400/>

The Fossify Phone description is simple and looks good enough to me and [Google Play](https://stackoverflow.com/questions/11071127/google-play-app-description-formatting) respects this formatting too. You might notice that in my version, there's no space between the short description and full description because I believe that should really be handled in the website layout. Speaking of spaces, I think the website would look nicer if the icons had some margin and maybe slightly rounded corners. Thoughts?

**Reply by IzzySoft** [reply]
> Actually it's still on my TODO list, I just never got around to it.

Ah, OK. Then I'll try to be more patient :wink: I've a job running once a month performing quality checks, which regularly reminds me that I don't pull the fulldesc for Gallery as I reworked it locally, which is what brings up that question every month again :see_no_evil: 

> You say Google Play will strip down the html tags?

I've stopped using PlayStore (and all the GMS stuff, going Goolge-free on my devices), am no Android dev myself and thus never put any app there – so I cannot talk of my own experience here, just repeat what I cobbled together from different sources. You can find details on this in the [IzzyOnDroid Fastlane Documentation](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane), more precisely [in this section](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane#formatting-descriptions). Google itself is not clear on this, saying "hop" in one of their docs and "ho" in another (e.g. on whether `<pre>` is supported or not). So my conclusion from those different sources was:

> Summing up tags permitted on Play (with some confidence) from above research: a href, b|i|u, h1|h2|h3, font color, br, small|sub|sup|blockquote, HTML entities, Unicode.

All of that is supported by IzzyOnDroid as well (with the exception of h1|h2|h3 being converted into p with the attributes bold and a bigger font size), and AFAIK F-Droid.org too. IoD (and F-Droid?) also supports ul|ol and code, for example. To make sure for F-Droid, you can check with [HTML descriptions](https://f-droid.org/en/docs/All_About_Descriptions_Graphics_and_Screenshots/#html-descriptions) there. For IoD, you can refer to the [`sanitizeHTML()` method in the `upstream.class.php`](https://gitlab.com/IzzyOnDroid/repo/-/blob/master/lib/upstream.class.php?ref_type=heads#L964) on full details.

Btw: Using multi-line HTML (or pseudo-HTML with `<br>`) I'd be very careful concerning F-Droid.org, as with their tools each newline is replaced by a `<br>` – so you'll have a lot of additional empty lines you'd probably not intend. At IoD that is taken care for by "compressing" the resulting HTML to a single line when importing `full_description.txt`.

> I edited the html on the webpage to draw a comparison between `<dl>`, `<dt>` formatting

Yeah, that was rather trying out things and not the "final variant", as I was waiting on what you planned to implement. I'd rather go with `<ul><li>` there, without the bullet points (replacing them with the "icons" using the `style` attribute in `<li>`). But then again, not sure how F-Droid and PlayStore would deal with that. Would be great to have someone trying out all those formattings with some not-so-important app listed on both places, so we could tell. I'd then update the Wiki page accordingly. (Btw: I was not able to show your screenshot full-size as it's embedded from some "private URL" – but I guess I got the idea without being able to read the text :see_no_evil:)

> no space between the short description and full description because I believe that should really be handled in the website layout.

That is handled in the website layout, yes. Which might not be perfect, I admit – I'm no web designer…

> I think the website would look nicer if the icons had some margin and maybe slightly rounded corners. Thoughts?

Suggestions appreciated! If you hand me the corresponding CSS I can take a look at that. Maybe like this:

```css
#desc .appicon {
  border-radius: .5em;
  margin-right: .5em;
}
```

Would that match your suggestion? LGTM:

![image](https://github.com/FossifyOrg/Phone/assets/6781438/e7be6b35-3937-4fe6-8599-7367f5b6c468)

Now it would look better were the description HTML. Let me make the first paragraph a `<p>`:

![image](https://github.com/FossifyOrg/Phone/assets/6781438/19894692-e69a-4dd8-88a9-24ba1f0e3f4e)

See what I mean? Using `nl2br` instead of converting to proper HTML has its limits. Unfortunately I cannot even use my Markdown parser to convert it. I mean, I could if you want – but it would then turn each section into a single paragraph. No problem with the "pseudo-bullet-points" <s>maybe</s> at all (yay, two trailing spaces, perfectly fits Markdown for a line break!), but with the [URL list](https://github.com/FossifyOrg/Phone/blob/master/fastlane/metadata/android/en-US/full_description.txt#L20) as that would just become a "wall-of-text" then. Prepend each line there with an asterisk+space (`* `), and markdown would convert it into a bullet-point list. Optionally, make the upper-case "headings" bold (e.g. `📱 <b>YOUR PRIVACY, OUR PRIORITY:</b>  `), and the results should be fine. Give it a try for Phone? Just manually push a modified `full_description.txt`, let me know when it's up, and I'll manually trigger a fastlane update. Should it look fine we can leave it at that, otherwise roll back for now. That would be an easy minimal adjustment bringing us to 95% of the target I guess – if not even to full 100%, depending on how it looks. And also easy to handle for your Weblate contributors. Thoughts? :wink:

**Reply by naveensingh** [reply]
That is a lot of thoughts ;) 

> Would be great to have someone trying out all those formattings with some not-so-important app listed on both places, so we could tell.

Keeping your suggestions in mind, I'll upload a formatted description with Fossify Keyboard (to be released soon-ish) and we can iterate with that. If it works well, I'll migrate the format to all other apps.

### IzzyOnDroid

> #desc .appicon {
  border-radius: .5em;
  margin-right: .5em;
}
> Would that match your suggestion?

Yep, this is exactly what I meant. To remedy the alignment issues, I'd do something like the following:

<img src="https://github.com/FossifyOrg/Phone/assets/36371707/436218ea-b88f-4960-aca1-75937010ac69" width=750/><br>
This is how [F-Droid](https://f-droid.org/en/packages/org.fossify.phone/) does it (using [flexbox](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_flexible_box_layout/Basic_concepts_of_flexbox)). I like this version but if you'd rather keep the app name centered, the app icon could be positioned above the title:

<img src="https://github.com/FossifyOrg/Phone/assets/36371707/0fddf89c-53fa-4093-995c-d1b5ae944ed6" width=750/><br>
Of course, these are just suggestions for the website in general and the current layout is perfectly fine. 
Thanks for the info and the links!

**Reply by IzzySoft** [reply]
Also nice ideas, thanks! But that would mean switching the whole design (template etc), which I'd rather not (at this point). A rewrite of the web part ("the repo browser") is planned to make it FOSS and available to be used by others, that's where this would come in handy. I hope to remember it when it comes to that (eh, adding it at the proper place of my todo list shall take care for that – done!). For now I'd go with the CSS from above, slightly modified:

```css
#desc .appicon {
  border-radius: .5em;
  margin: .5em .5em .5em 0;
}
#desc .appicon + p { margin-top: 7px; } /* adjust for app-icon with HTML descriptions */
```

Fossify Phone (plain text description):
![image](https://github.com/FossifyOrg/Phone/assets/6781438/f8125f2d-6fcf-4fa0-87df-702183ea477e)

Fossify Gallery (HTML description):
![image](https://github.com/FossifyOrg/Phone/assets/6781438/da8b26d8-81cc-4b1e-9324-5e76b27939e6)

Best compromise I could come up with – as there's no CSS selector for "plain text (not an element)" :man_shrugging:

But I agree, the icon would better fit in front of name+short-desc. Will look at that later (maybe as much later as with the rewrite). Thanks a lot for the suggestion, Naveen!
